File: src/main/services/ai/claude-client.ts
import Anthropic from '@anthropic-ai/sdk';
import { ActionItem, NormalizedEvent } from '../../../shared/types/calendar';
import { tokenStore } from '../auth/token-store';

let clientInstance: Anthropic | null = null;

async function getClient(): Promise<Anthropic> {
  if (clientInstance) return clientInstance;

  const apiKey = await tokenStore.loadSecret('claude-api-key');
  if (!apiKey) {
    throw new Error('Claude API key not configured. Please add it in Settings.');
  }

  clientInstance = new Anthropic({ apiKey });
  return clientInstance;
}

export function resetClaudeClient(): void {
  clientInstance = null;
}

export async function extractActionItems(
  note: string,
  event: NormalizedEvent
): Promise<ActionItem[]> {
  const client = await getClient();

  const start = new Date(event.start);
  const end = new Date(event.end);
  const durationMinutes = Math.round((end.getTime() - start.getTime()) / 60000);

  const prompt = `You are an assistant that extracts action items from meeting notes.

Meeting: "${event.title}"
Date: ${start.toLocaleDateString('en-US', { weekday: 'long', year: 'numeric', month: 'long', day: 'numeric' })}
Duration: ${durationMinutes} minutes
Attendees: ${event.attendees.join(', ') || 'Not specified'}

Meeting notes:
${note}

Extract all action items from these notes. Return ONLY a JSON object with this exact structure:
{
  "actionItems": [
    {
      "id": "1",
      "title": "short actionable title (verb + object)",
      "description": "additional context or null",
      "dueDate": "YYYY-MM-DD or null",
      "assignee": "name or email if mentioned, or null",
      "priority": "urgent | important | normal"
    }
  ]
}

Rules:
- Only extract explicit commitments, tasks, and follow-ups — not general discussion points
- Titles must be action-oriented (start with a verb: "Send", "Schedule", "Review", "Create", etc.)
- If no action items, return {"actionItems": []}
- Infer due dates only when clearly stated ("by Friday", "next Monday", "EOD", "end of week")
- Today is ${start.toISOString().split('T')[0]}
- Set priority "urgent" only for explicit deadlines within 24 hours or words like "ASAP", "urgent", "critical"
- Set priority "important" for items marked important or needed soon (this week)
- Default to "normal" for everything else
- Return valid JSON only, no markdown, no surrounding text`;

  const message = await client.messages.create({
    model: 'claude-opus-4-5',
    max_tokens: 1024,
    messages: [{ role: 'user', content: prompt }],
  });

  const content = message.content[0];
  if (content.type !== 'text') {
    throw new Error('Unexpected response type from Claude');
  }

  // Strip markdown code fences if the model wraps the response despite instructions
  const raw = content.text.replace(/^```(?:json)?\s*/i, '').replace(/\s*```\s*$/i, '').trim();
  const parsed = JSON.parse(raw);
  const items: ActionItem[] = parsed.actionItems ?? [];

  // Ensure IDs are unique
  return items.map((item, idx) => ({
    ...item,
    id: item.id ?? String(idx + 1),
  }));
}

File: lib/services/ai/claude_client.dart
import 'package:dio/dio.dart';
import '../../models/calendar_event.dart';
import '../../models/settings.dart';
import 'task_extractor.dart';

class ClaudeClient implements TaskExtractor {
  final Dio _dio = Dio(BaseOptions(
    baseUrl: 'https://api.anthropic.com',
    connectTimeout: const Duration(seconds: 10),
    receiveTimeout: const Duration(seconds: 60),
    headers: {
      'Content-Type': 'application/json',
      'anthropic-version': '2023-06-01',
    },
  ));

  String? _apiKey;
  String _modelId = kDefaultClaudeModelId;

  void setApiKey(String key) {
    _apiKey = key;
  }

  void setModel(String modelId) {
    _modelId = modelId;
  }

  @override
  Future<List<ActionItem>> extractActionItems(
    NormalizedEvent event, {
    String? transcript,
    String? summary,
    String? notes,
  }) async {
    if (_apiKey == null || _apiKey!.isEmpty) {
      throw Exception('Claude API key not configured');
    }

    final prompt = ExtractionPrompts.buildExtractPrompt(
      event,
      transcript: transcript,
      summary: summary,
      notes: notes,
    );

    try {
      final response = await _dio.post(
        '/v1/messages',
        options: Options(headers: {'x-api-key': _apiKey}),
        data: {
          'model': _modelId,
          'max_tokens': 1024,
          'messages': [
            {'role': 'user', 'content': prompt}
          ],
        },
      );

      final content = response.data['content'][0]['text'] as String;
      return ExtractionPrompts.parseActionItems(content);
    } on DioException catch (e) {
      throw _dioError(e);
    }
  }

  /// Generates a summary and extracts action items from a meeting transcript.
  @override
  Future<({String summary, List<ActionItem> actionItems})> summarizeTranscript(
    String transcript,
    NormalizedEvent event,
  ) async {
    if (_apiKey == null || _apiKey!.isEmpty) {
      throw Exception('Claude API key not configured');
    }

    final prompt = ExtractionPrompts.buildSummarizePrompt(transcript, event);

    try {
      final response = await _dio.post(
        '/v1/messages',
        options: Options(headers: {'x-api-key': _apiKey}),
        data: {
          'model': _modelId,
          'max_tokens': 2048,
          'messages': [
            {'role': 'user', 'content': prompt}
          ],
        },
      );

      final content = response.data['content'][0]['text'] as String;
      return ExtractionPrompts.parseSummary(content);
    } on DioException catch (e) {
      throw _dioError(e);
    }
  }

  Exception _dioError(DioException e) {
    final status = e.response?.statusCode;
    if (status == 404) return Exception('Model "$_modelId" not found (404). Please select a different model in Settings.');
    if (status == 401) return Exception('Invalid API key (401). Please check your Claude API key in Settings.');
    if (status == 429) return Exception('Rate limit exceeded (429). Please wait a moment and try again.');
    return Exception('API error $status: ${e.message}');
  }
}
